### ISSUE TITLE
Project Plan: House Price Predictor

### ISSUE THREAD

**Issue by SergeATX** [TriangulumAustrale/Scikit-learn#2]
# House Price Predictor — Project Plan

## 1. Purpose

A personal portfolio project for Antony Tatarenko (recent CS grad, B.S. May
2026, Texas State University), built independently of his work at machthirty.
It applies coursework in regression and feature importance to a small,
real, end-to-end ML product: a web app that predicts California house prices
from a handful of input features and explains which features drove the
prediction.

Kept deliberately small and finishable — the goal is a working, deployed,
explainable product with clean commit/issue history, not a research project
or a maximal feature set.

## 2. What it does

1. User enters a small set of house features (median income, house age,
   average rooms, average bedrooms, population, average occupancy, latitude,
   longitude) in a web form.
2. Backend loads a pre-trained `RandomForestRegressor` and returns a
   predicted median house value.
3. Response includes the top 3 features (by importance) that most influenced
   that prediction, in plain English.
4. Frontend displays the predicted price and the contributing-features list.

## 3. Scope constraints

- **One dataset**: `sklearn.datasets.fetch_california_housing()`. No external
  data sourcing, scraping, or cleaning pipeline — this dataset ships ready to
  use inside scikit-learn.
- **One model**: `RandomForestRegressor`. Do not add model comparison,
  hyperparameter search infrastructure, or ensembling — a single, explainable
  model that Antony can describe line-by-line in an interview matters more
  than squeezing out marginal accuracy.
- **No persistence/auth in v1**: stateless request → predict → respond flow.
  No database, no user accounts, no saved prediction history.
- **No retraining pipeline in v1**: the model is trained once (offline, via a
  script) and the artifact is loaded at API startup. Live retraining, model
  versioning, and MLOps tooling are explicitly out of scope.

## 4. Tech stack

Matches the stack used on the machthirty log-anomaly-detector project, to
keep marginal learning cost low:

| Layer | Choice | Why |
|---|---|---|
| Backend | FastAPI (Python) | Async-friendly, typed, pairs naturally with scikit-learn |
| ML | scikit-learn `RandomForestRegressor` | Explainable, built-in `.feature_importances_`, no extra deps |
| Model artifact | `joblib` | Standard scikit-learn model serialization |
| Frontend | Vite + React | Current industry default; CRA is deprecated |
| UI library | MUI | Ready-made form/card/alert components, avoids hand-rolled CSS time sink |
| Backend tests | pytest | Covers model loading, prediction endpoint, edge cases |
| Deployment (backend) | AWS Lambda + API Gateway (via Mangum) | Serverless, free-tier friendly, strong "cloud-native" resume signal |
| Deployment (frontend) | AWS S3 (static site) | Simplest static hosting on free tier |

## 5. Architecture

```
┌─────────────┐      HTTP       ┌──────────────────┐      loads       ┌──────────────┐
│  Frontend    │ ─────────────▶ │  FastAPI backend  │ ───────────────▶ │ model.joblib │
│ (Vite+React) │ ◀───────────── │  (Lambda via      │                  │  (artifact)  │
│  S3 static   │   JSON result  │   Mangum adapter)  │                  └──────────────┘
└─────────────┘                 └──────────────────┘
```

- Model is trained offline by a standalone script (`backend/train.py`),
  producing `backend/model/model.joblib` and
  `backend/model/feature_importances.json`.
- The FastAPI app loads the artifact once at cold start and serves
  predictions from memory — no per-request retraining or disk I/O beyond
  startup.
- No database anywhere in the request path.

## 6. Data & model

**Dataset**: California Housing (`fetch_california_housing`), 20,640 rows,
8 numeric features, target = median house value (in $100,000s).

**Features** (all numeric, no preprocessing/encoding needed):
`MedInc`, `HouseAge`, `AveRooms`, `AveBedrms`, `Population`, `AveOccup`,
`Latitude`, `Longitude`.

**Model**: `RandomForestRegressor` (scikit-learn defaults, minor tuning
allowed — e.g. `n_estimators=100`, `max_depth` capped for speed — but no
hyperparameter search infrastructure).

**Evaluation**: report R² and MAE on a held-out test split (80/20). No formal
target threshold — the point is demonstrating a working, evaluated model, not
chasing a specific score.

**Explainability**: `model.feature_importances_` computed once at training
time and saved alongside the model. At prediction time, the API reports the
3 globally-most-important features (not per-prediction SHAP-style
explanations — that's out of scope for v1 and harder to justify in an
interview at this stage).

### `POST /predict`

Request body:
```json
{
  "MedInc": 8.3,
  "HouseAge": 25,
  "AveRooms": 6.2,
  "AveBedrms": 1.1,
  "Population": 1500,
  "AveOccup": 3.0,
  "Latitude": 34.05,
  "Longitude": -118.25
}
```

Response body:
```json
{
  "predicted_value": 4.52,
  "predicted_value_usd": 452000,
  "top_features": [
    {"feature": "MedInc", "importance": 0.52},
    {"feature": "AveOccup", "importance": 0.14},
    {"feature": "Longitude", "importance": 0.09}
  ]
}
```

- All 8 fields required, validated via a Pydantic model (numeric, sane
  ranges — e.g. `Latitude` between 32 and 42, matching California).
- Invalid/out-of-range input → `422` with a clear validation error message.

### `GET /health`

Returns `{"status": "ok"}` — used for deployment smoke-testing and as the
scaffolding acceptance check.

## 8. Frontend / UX plan

Single-screen app, no routing needed:

1. **Input form** — 8 numeric fields (MUI `TextField`s) with labels in plain
   English (e.g. "Median Income (10k USD)" instead of `MedInc`), reasonable
   default/placeholder values pre-filled so the form is usable without
   external documentation.
2. **Submit** — calls `POST /predict`, shows a loading state (MUI
   `CircularProgress`) while waiting.
3. **Result card** — predicted price (large, prominent) + a small list of
   the top 3 contributing features in plain English (e.g. "Median income in
   this area was the biggest driver of this estimate").
4. **Error state** — validation errors from the API (422) or network
   failures shown via MUI `Alert`, without crashing the form.

No charts/graphs in v1 — consistent with the machthirty project's "no
charts required" scoping decision; a single result card is sufficient here.

## 9. Testing strategy

- **Backend (pytest)**:
  - `train.py` produces a model artifact and a feature-importances file
    that both load correctly.
  - `/predict` returns a well-formed response for valid input.
  - `/predict` returns `422` for missing/out-of-range input.
  - `/health` returns `{"status": "ok"}`.
- **Frontend**: manual verification only in v1 (matches the machthirty
  project's scoping — automated frontend tests are lowest-priority relative
  to shipping the ML/API/deployment work that carries the actual resume
  signal).

## 10. Deployment plan

- **Backend**: package FastAPI app with Mangum adapter, deploy to Lambda,
  expose via API Gateway (HTTP API, not REST API, for lower cost/complexity).
  Model artifact bundled into the deployment package (small enough — no S3
  model-loading needed).
- **Frontend**: `npm run build`, upload static output to an S3 bucket
  configured for static website hosting. CloudFront is optional/stretch, not
  required for a working public URL.
- **Environment config**: frontend's API base URL set via a build-time env
  var (`VITE_API_URL`), not hardcoded.

## 11. Milestones (drive GitHub issues 1:1)

0. **Project scaffolding** — FastAPI + Vite/React/MUI skeleton. *(Issue #1,
   already created.)*
1. **Data & model training** — `train.py` script: load dataset, train/test
   split, train `RandomForestRegressor`, evaluate, save model + feature
   importances.
2. **Backend prediction API** — `/predict` and `/health` endpoints, Pydantic
   request/response schemas, loads the saved model artifact.
3. **Backend tests** — pytest suite covering training output, `/predict`,
   `/health`, and validation error cases.
4. **Frontend UI** — input form + result card + error/loading states, wired
   to the live API.
5. **Deployment** — backend to Lambda + API Gateway, frontend to S3, both
   reachable at a real public URL.
6. **README** — what it does, how to run locally, architecture, why
   RandomForestRegressor was chosen, what the feature importances mean.

## 12. Definition of done (v1)

- Live public URL where anyone can enter house features and get a prediction
  with a plain-English explanation.
- Backend has passing pytest coverage for the model and API layers.
- README documents the project accurately for both a recruiter skim and an
  interview deep-dive.
- No known scope creep beyond what's listed above (single dataset, single
  model, no auth/persistence/retraining pipeline).

## 13. Explicitly out of scope (future stretch, not v1)

- Model comparison / hyperparameter search / MLOps pipeline.
- Per-prediction explainability (SHAP, LIME).
- Saved prediction history / user accounts / database.
- CloudFront CDN in front of the S3 frontend.
- Automated frontend tests.
